proof/refine/ARM/Reply_R.thy

```diff
@@ -1081,27 +1024,841 @@ lemma sc_replies_relation_replyNext_update:
+                                                                                    - minSchedContextBits)))
+                              and K (rp = hd (sc_replies sc))"
+                      and P'="ko_at' sc' ptr"  in corres_inst)
+          apply (rule corres_gen_asm')
+          apply (rule corres_guard_imp)
+            apply (rule_tac sc=sc and sc'=sc' in setSchedContext_update_corres; simp?)
+             apply (clarsimp simp: sc_relation_def objBits_simps)+
+         apply (wpsimp wp: get_sched_context_exs_valid simp: is_sc_obj_def obj_at_def)
+          apply (rename_tac ko; case_tac ko; clarsimp)
+         apply simp
+        apply (wpsimp simp: obj_at_def is_sc_obj_def vs_heap_simps)
+       apply (wpsimp wp: get_sched_context_no_fail)
+      apply (clarsimp simp: obj_at_def is_sc_obj_def)
+     apply simp
+    apply wpsimp+
+  done
+
+lemma replyPrevNext_update_commute:
+  "replyPrev_update Map.empty (replyNext_update Map.empty reply)
+   = replyNext_update Map.empty (replyPrev_update Map.empty reply)"
+  by (cases reply; clarsimp)
+
+lemma updateReply_Prev_Next_rewrite:
+  "monadic_rewrite False True (reply_at' rp)
+    (do y <- updateReply rp (replyPrev_update Map.empty);
+         updateReply rp (replyNext_update Map.empty)
+     od)
+    (do y <- updateReply rp (replyNext_update Map.empty);
+         updateReply rp (replyPrev_update Map.empty)
+     od)"
+  apply (clarsimp simp: monadic_rewrite_def)
+  apply (rule monad_state_eqI)
+    apply (find_goal \<open>match conclusion in "_ = _" \<Rightarrow> -\<close>)
+    subgoal
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyPrev_update Map.empty" in meta_spec2)
+    apply (insert no_fail_updateReply, drule_tac x=rp and y="replyNext_update Map.empty" in meta_spec2)
+    apply (rule; clarsimp simp: in_monad no_fail_def snd_bind split_def)
+     apply (drule (1) use_valid[OF _ updateReply_reply_at'], fastforce)+
+    done
+   apply (all \<open>clarsimp simp: updateReply_def getReply_def setReply_def getObject_def2
+                              obj_at'_def projectKOs updateObject_default_def setObject_def
+                              loadObject_default_def2[simplified] ARM_H.fromPPtr_def ps_clear_upd'
+                              split_def in_monad in_magnitude_check' objBits_simps'\<close>)
+   apply (fastforce simp add: fun_upd_def replyPrevNext_update_commute)+
+  done
+
+lemma reply_sc_update_sc_with_reply_None:
+  "set_reply_obj_ref reply_sc_update rp None \<lbrace>\<lambda>s. sc_with_reply rp s = None\<rbrace>"
+  apply (wpsimp wp: update_sk_obj_ref_wp)
+  apply (drule sc_with_reply_NoneD)
+  apply (simp add: sc_with_reply_def sc_replies_sc_at_def obj_at_def)
+  apply (rule the_pred_option_None)
+  by (metis Structures_A.kernel_object.simps(45) option.simps(1))
+
+(* cleanReply corres when rp is not in a reply stack *)
+lemma cleanReply_sc_with_reply_None_corres:
```

Comment:
Since we have `cleanReply_sr_inv` earlier. It seems as if this is effectively comparing `set_reply_obj_ref reply_sc_update rp None` with `return ()`? If this lemma is really quite useful in the context it's used, I suggest renaming it or adding a comment to remind the ready that `cleanReply` is really a no-op here.

Reply:
Right, yes, it is true that this has no-op on both sides, and we can do a bit more abstraction here. Let me see...

Reply:
Ok, I think I now remember why this is done in this particular way. I could easily write a version of this lemma using`cleanReply_sr_inv` and a `sr_inv` version of `corres_noop2`, but it will require `sym_refs (state_refs_of s)` in the assumption and it is really a pain to carry it through in the locations where this lemma is used in the main corres proof. I think I will stick to the current version and add some comments.